compiler/rustc_middle/src/ty/layout.rs

```diff
@@ -3482,3 +3482,154 @@ fn make_thin_self_ptr<'tcx>(
+/// * zero init enums where a discriminant with tag 0 exists, but is invalid to be zeroed
+/// * zero init type that does not allow zero init (only inside arrays)
+///
+/// A strict form of these checks that uses const evaluation exists in
+/// `rustc_const_eval::might_permit_raw_init`, and a tracking issue for making these checks
+/// stricter is <https://github.com/rust-lang/rust/issues/66151>.
+///
+/// FIXME: Once all the conservatism is removed from here, and the checks are ran by default,
+/// we can use the const evaluation checks always instead.
+pub fn might_permit_raw_init<'tcx, C>(this: TyAndLayout<'tcx>, cx: &C, init_kind: InitKind) -> bool
+where
+    C: HasDataLayout + ty::layout::HasParamEnv<'tcx> + ty::layout::HasTyCtxt<'tcx>,
+{
+    return might_permit_raw_init_inner(this, cx, init_kind, false);
+
+    fn might_permit_raw_init_inner<'tcx, C>(
+        this: TyAndLayout<'tcx>,
+        cx: &C,
+        init_kind: InitKind,
+        inside_array: bool,
+    ) -> bool
+    where
+        C: HasDataLayout + ty::layout::HasParamEnv<'tcx> + ty::layout::HasTyCtxt<'tcx>,
+    {
+        let scalar_allows_raw_init = move |s: Scalar| -> bool {
+            match init_kind {
+                InitKind::Zero => {
+                    // The range must contain 0.
+                    s.valid_range(cx).contains(0)
+                }
+                InitKind::Uninit => {
+                    // FIXME(#66151) This should be "is the type a union", but that's pending on a
+                    // resolution to https://github.com/rust-lang/unsafe-code-guidelines/issues/71
+                    // And likely a large number of crates fail with those rules, though no crater
+                    // run has been done yet.
+                    //
+                    // The range must include all values.
+                    s.is_always_valid(cx)
+                }
+            }
+        };
+
+        // These are bypasses in order to try not to break quite as many crates so quickly.
+        // They are being done only if we're inside an array, to ensure we don't panic on *less*
+        // things than we did before this change.
+        // See: https://github.com/rust-lang/rust/pull/99389
+        if inside_array {
+            match init_kind {
+                // We panic if creating this type with all 0x01 bytes would
+                // cause LLVM UB.
+                //
+                // Therefore, in order for us to not panic, it must either be a
+                // reference to [T] where T has align 1 (where we don't statically know
+                // the size, so we don't emit any dereferenceable), or a reference to str
+                // which acts much like a [u8].
+                //
+                // We *do* need to panic for &dyn Trait, even though the layout of dyn Trait is
+                // size 0 align 1, because &dyn Trait holds a reference to a non-zero sized type,
+                // which also must be aligned.
```

Comment:
```suggestion
                // size 0 align 1, because &dyn Trait holds a reference to the vtable, which
                // must be aligned and dereferenceable.
```